src/Compilers/Core/Portable/FileSystem/PathUtilities.cs

```diff
@@ -615,6 +616,41 @@ private static int PathHashCode(string path)
             return hc;
         }
 
+        public static string NormalizePathPrefix(string filePath, ImmutableArray<KeyValuePair<string, string>> pathMap)
+        {
+            if (pathMap.IsDefaultOrEmpty)
+            {
+                return filePath;
+            }
+
+            // find the first key in the path map that matches a prefix of the normalized path (followed by a path separator).
+            // Note that we expect the client to use consistent capitalization; we use ordinal (case-sensitive) comparisons.
+            foreach (var kv in pathMap)
+            {
+                var oldPrefix = kv.Key;
+                if (!(oldPrefix?.Length > 0)) continue;
+
+                if (filePath.StartsWith(oldPrefix, StringComparison.Ordinal) && filePath.Length > oldPrefix.Length && PathUtilities.IsDirectorySeparator(filePath[oldPrefix.Length]))
```

Comment:
>PathUtilities.IsDirectorySeparator [](start = 118, length = 34)

Just realized -- this changes semantics on Unix. The helper in PathUtilities uses the current platform separator(s), but here we always want both / and \